### PR TITLE
Refactor and Build

### PR DESCRIPTION
**Pull request by forestbat**
refactor:refactor ShaderManager methods and fields to static 
build:update gradle to 5.4.1

### CODE REVIEW

File: EngineAPI/src/main/java/nullengine/event/world/item/ItemThrowEvent.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import nullengine.event.Event;
+import nullengine.item.Item;
+import nullengine.world.World;
+
+public class ItemThrowEvent implements Event {
+	private World world;
+	private Entity player;
+	private Item item;
+	private Block block;
+	public ItemThrowEvent(Item item,World world, Entity player, Block block){
```
Comment (mouse0w0) [inline EngineAPI/src/main/java/nullengine/event/world/item/ItemThrowEvent.java:14]: Unnecessary `Block` parameter, should add `ItemEntity` parameter.

File: EngineAPI/src/main/java/nullengine/event/world/item/ItemBreakEvent.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+	private Item item;
+	public ItemBreakEvent(Item item,World world, Entity player){
+		this.world=world;
+		this.item=item;
+		this.player=player;
+	}
+
+	@Override
+	public boolean isCancellable() {
+		return true;
```
Comment (mouse0w0) [inline EngineAPI/src/main/java/nullengine/event/world/item/ItemBreakEvent.java:20]: `isCancellable` return true don't mean this event is cancellable. Please implement interface `Cancellable`.

File: EngineAPI/src/main/java/nullengine/event/world/item/ItemThrowEvent.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+		this.world=world;
+		this.item=item;
+		this.player=player;
+	}
+
+	@Override
+	public boolean isCancellable() {
+		return true;
```
Comment (mouse0w0) [inline EngineAPI/src/main/java/nullengine/event/world/item/ItemThrowEvent.java:22]: Please implement `Cancellable`

**Review by mouse0w0 — COMMENTED** [review]
Many problems need to solve.

File: Engine/src/main/java/nullengine/game/GameBase.java [hunk 4]
```diff
@@ -29,13 +29,13 @@ public GameBase(Engine engine) {
     }
 
     /**
-     * Construct stage, collect mod and resource according to it option
+     * Construct stage, collect mod and nullengine.resource according to it option
```
Comment (mouse0w0) [inline Engine/src/main/java/nullengine/game/GameBase.java]: This is an incorrect change.

File: Engine/src/main/java/nullengine/game/GameBase.java [hunk 5]
```diff
@@ -29,13 +29,13 @@ public GameBase(Engine engine) {
...
      */
     protected void constructStage() {
     }
 
     /**
-     * let mod and resource related module load resources.
+     * let mod and nullengine.resource related module load resources.
```
Comment (mouse0w0) [inline Engine/src/main/java/nullengine/game/GameBase.java]: This is an incorrect change.

File: Engine/src/main/java/nullengine/game/GameBase.java [hunk 6]
```diff
@@ -95,5 +95,5 @@ public boolean isTerminated() {
         return terminated;
     }
 
-    // TODO: unload mod/resource
+    // TODO: unload mod/nullengine.resource
```
Comment (mouse0w0) [inline Engine/src/main/java/nullengine/game/GameBase.java]: This is an incorrect change.

File: Engine/src/main/java/nullengine/game/GameServerFullAsync.java [hunk 7]
```diff
@@ -74,7 +74,7 @@ protected void tryTerminate() {
         for (WorldCommon worldCommon : internalWorlds) {
             worldCommon.stop();
         }
-        // TODO: unload mod/resource here
+        // TODO: unload mod/nullengine.resource here
```
Comment (mouse0w0) [inline Engine/src/main/java/nullengine/game/GameServerFullAsync.java]: This is an incorrect change.

File: EngineClient/src/main/java/nullengine/client/rendering/util/GLHelper.java [hunk 8]
```diff
@@ -71,7 +71,7 @@ public static ByteBuffer getResourcesAsBuffer(String resource, int size) throws
                     }
                 }
             } catch (Exception e) {
-                throw new IOException(String.format("cannot load resource: %s", resource), e);
+                throw new IOException(String.format("cannot load nullengine.resource: %s", resource), e);
```
Comment (mouse0w0) [inline EngineClient/src/main/java/nullengine/client/rendering/util/GLHelper.java]: This is an incorrect change.

**Review by mouse0w0 — COMMENTED** [review]
I don't know what's the use of `resource`. If you want to get resources. Please see `AssetManager` and `ModAssets`.

File: Engine/src/main/java/nullengine/world/WorldCommon.java [hunk 9]
```diff
@@ -71,11 +73,11 @@ public void spawnEntity(Entity entity) {
         entityList.add(entity);
     }
 
-    @Deprecated
-    public void playerJoin(Player player) {
-        // FIXME:
+
+    public void onPlayerJoin(@Nonnull Player player) {
```
Comment (mouse0w0) [inline Engine/src/main/java/nullengine/world/WorldCommon.java]: Don't rename this method.

File: Engine/src/main/java/nullengine/world/WorldCommon.java [hunk 10]
```diff
@@ -71,11 +73,11 @@ public void spawnEntity(Entity entity) {
...
+        Player player1=Validate.notNull(player);
```
Comment (mouse0w0) [inline Engine/src/main/java/nullengine/world/WorldCommon.java]: Please use `Objects.requireNonNull`.

You don't need to receive method return value. For example:
```java
public void playerJoin(@Nonnull Player player) {
        Objects.requireNonNull(player); // if player is null, will throw NullPointerException.
        var entity = new PlayerEntity(entityList.size(), this);
        player.controlEntity(entity);
        player.controlEntity(entity);
        spawnEntity(entity);
        players.add(player);
}
```

**Review by mouse0w0 — COMMENTED** [review]
Don't remove `ShaderManager.INSTANCE` and don't change methods to static.

**Review by mouse0w0 — COMMENTED** [review]
Invaild revert commit message. Please see [CONTRIBUTING.md#Revert](https://github.com/UnknownDomainGame/NullEngine/blob/dev/CONTRIBUTING.md#revert).

For example: 
```
revert: add ItemImpl and two events

This reverts commit 9fdc950edc6600b518e47b7aed08e8db85acf101.
```

File: Engine/src/main/java/nullengine/item/ItemImpl.java [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+import nullengine.event.world.item.ItemThrowEvent;
+import nullengine.world.World;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+
+public class ItemImpl implements Item,Runnable {
```
Comment (ci010) [inline Engine/src/main/java/nullengine/item/ItemImpl.java]: I'm just curious. Why ItemImpl is a Runnable?

What is the design here? Am I missing any context about this? @Mouse0w0

File: Engine/src/main/java/nullengine/item/ItemImpl.java [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+		public ItemBuilder setMaxStack(int maxStack){
+			maxStackAmount=maxStack;
+			return this;
+		}
+		public ItemImpl build(){
+			return new ItemImpl(name,hp,harm,canBeDestroy,maxStackAmount);
+		}
+	}
+	public void onItemUse(ItemImpl item, World world, Entity playerEntity,int tick){
+
```
Comment (ci010) [inline Engine/src/main/java/nullengine/item/ItemImpl.java]: Maybe reference how the Block implements these. (Using `BlockBehavior`).

Looks like Item have this design too. Please see the file [EngineAPI/src/main/java/nullengine/item/component/UseBehavior.java](https://github.com/UnknownDomainGame/NullEngine/tree/dev/EngineAPI/src/main/java/nullengine/item/component)

Comment (forestbat) [reply Engine/src/main/java/nullengine/item/ItemImpl.java]: This class has been removed，and you can see the last two commits。

Comment (forestbat) [reply Engine/src/main/java/nullengine/item/ItemImpl.java]: 而且都是中国人半天不讲汉语.jpg

**Comment by mouse0w0** [comment]
我非常不建议在这个Pull Request上继续讨论下去了，已经是改来改去积重难返。。。

**Pull request closed without merging**